src/main/java/net/glowstone/block/data/IBlockData.java

```diff
@@ -0,0 +1,21 @@
+package net.glowstone.block.data;
+
+import net.glowstone.block.data.state.StateGenerator;
+import net.glowstone.block.data.state.StateValue;
+import org.bukkit.block.data.BlockData;
+
+import java.util.Set;
+
+public interface IBlockData extends BlockData {
+
+    Set<StateValue<?>> getStateValues();
+
+    default <T extends Object> StateValue<T> getStateValue(StateGenerator<T> generator){
+        return (StateValue<T>) this.getStateValues().stream().filter(s -> s.getGenerator().equals(generator)).findAny().orElse(null);
+    }
+
+    default StateValue<?> getStateValue(String id){
+        return this.getStateValues().stream().filter(s -> s.getGenerator().getId().equals(id)).findAny().orElse(null);
+    }
```

Comment:
Nitpick: Personally, I prefer using Optionals instead of nulls when a value may or may not be present, since it's easy to tell at a glance that this is the intended behavior. That's just me, though.

Reply:
Personally I also prefer to use Optionals too, but because the Bukkit API didnt, I chose not to in the name of keeping coherence. But I am willing to change it to optional in the name of pushing for modern API